### PR TITLE
Reduce Optional_string usage

### PR DESCRIPTION
**Pull request by jasondegraw**
Pull request overview
---------------------
Reduce the usage of the various forms of string optional arguments by about 50%. In many cases, the optional can be replaced with a `std::string_view` and the presence of the string can be checked with the `empty` method. Most of the remaining cases either require a deeper refactor or are too tied up with other ongoing refactors that there's not a good way to get at changing just the optional argument. In a few situations, overloading has been used to shift run-time choices of behavior to compile-time choices of behavior, so there may be a (very) small performance benefit from this PR.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/DXCoils.cc [hunk 1]
```diff
@@ -15687,8 +15687,8 @@ void GetDXCoilIndex(EnergyPlusData &state,
                     std::string const &DXCoilName,
                     int &DXCoilIndex,
                     bool &ErrorsFound,
-                    Optional_string_const ThisObjectType,
-                    Optional_bool_const SuppressWarning)
+                    std::string_view const ThisObjectType,
```
Comment (jasondegraw) [inline src/EnergyPlus/DXCoils.cc:15690]: This one is a good example of the easier changes. The argument is only used for errors, so the argument type is changed first.

File: src/EnergyPlus/DXCoils.cc [hunk 2]
```diff
@@ -15706,11 +15706,10 @@ void GetDXCoilIndex(EnergyPlusData &state,
 
     DXCoilIndex = UtilityRoutines::FindItemInList(DXCoilName, state.dataDXCoils->DXCoil);
     if (DXCoilIndex == 0) {
-        if (present(SuppressWarning)) {
+        if (!SuppressWarning) {
             //     No warning printed if only searching for the existence of a DX Coil
-        } else {
-            if (present(ThisObjectType)) {
-                ShowSevereError(state, ThisObjectType + ", GetDXCoilIndex: DX Coil not found=" + DXCoilName);
+            if (!ThisObjectType.empty()) {
```
Comment (jasondegraw) [inline src/EnergyPlus/DXCoils.cc:15711]: Then the call to `present` is changed to use the `empty` method.

File: src/EnergyPlus/DXCoils.cc [hunk 3]
```diff
@@ -15706,11 +15706,10 @@ void GetDXCoilIndex(EnergyPlusData &state,
...
+                ShowSevereError(state, fmt::format("{}, GetDXCoilIndex: DX Coil not found={}", ThisObjectType, DXCoilName));
```
Comment (jasondegraw) [inline src/EnergyPlus/DXCoils.cc:15712]: And finally the string handling is changed to something compatible with `string_view`.

File: src/EnergyPlus/NodeInputManager.cc [hunk 4]
```diff
@@ -867,13 +871,11 @@ void InitUniqueNodeCheck(EnergyPlusData &state, std::string const &ContextName)
     state.dataNodeInputMgr->CurCheckContextName = ContextName;
 }
 
-void CheckUniqueNodes(EnergyPlusData &state,
-                      std::string const &NodeTypes,
-                      std::string const &CheckType,
-                      bool &ErrorsFound,
-                      Optional_string_const CheckName,
-                      Optional_int_const CheckNumber,
-                      Optional_string_const ObjectName)
```
Comment (jasondegraw) [inline src/EnergyPlus/NodeInputManager.cc:876]: This function was checking for uniqueness of an optional name or optional number based on the string in CheckType... which appears to always be one of two literal strings... so known at runtime. Replace that with two functions, one that checks for name and one that checks for number.

File: src/EnergyPlus/DXCoils.cc [hunk 5]
```diff
@@ -15720,7 +15719,7 @@ void GetDXCoilIndex(EnergyPlusData &state,
 }
 
 std::string
-GetDXCoilName(EnergyPlusData &state, int &DXCoilIndex, bool &ErrorsFound, Optional_string_const ThisObjectType, Optional_bool_const SuppressWarning)
+GetDXCoilName(EnergyPlusData &state, int &DXCoilIndex, bool &ErrorsFound, std::string_view const ThisObjectType, bool const SuppressWarning)
```
Comment (amirroth) [inline src/EnergyPlus/DXCoils.cc:15722]: Functions like this should not be necessary. This should just be `state.dataDXCoils->DXCoils(DXCoilNum).Name`. The error checking and defaulting should not be necessary. You should not need to call a function to grab a field of an object whose index is known. I understand that this has nothing to do with this PR, but this is as good a place as any for me to hang this comment.

File: src/EnergyPlus/DXCoils.hh [hunk 6]
```diff
@@ -708,11 +708,11 @@ namespace DXCoils {
...
-                        Optional_string_const ThisObjectType,
-                        Optional_bool_const SuppressWarning);
+                        std::string_view const ThisObjectType = {},
+                        bool const SuppressWarning = false);
 
     std::string GetDXCoilName(
-        EnergyPlusData &state, int &DXCoilIndex, bool &ErrorsFound, Optional_string_const ThisObjectType, Optional_bool_const SuppressWarning);
+        EnergyPlusData &state, int &DXCoilIndex, bool &ErrorsFound, std::string_view const ThisObjectType = {}, bool const SuppressWarning = false);
 
     Real64 GetCoilCapacity(EnergyPlusData &state,
```
Comment (amirroth) [inline src/EnergyPlus/DXCoils.hh:717]: Same thing.

File: src/EnergyPlus/DataSurfaceColors.hh [hunk 7]
```diff
@@ -92,7 +92,7 @@ namespace DataSurfaceColors {
                                     std::string const &ColorType // for now, must be DXF
     );
 
-    void SetUpSchemeColors(EnergyPlusData &state, std::string const &SchemeName, Optional_string_const ColorType = _);
+    void SetUpSchemeColors(EnergyPlusData &state, std::string const &SchemeName, std::string const &ColorType);
```
Comment (amirroth) [inline src/EnergyPlus/DataSurfaceColors.hh]: `std::string_view`

File: src/EnergyPlus/InputProcessing/InputProcessor.cc [hunk 8]
```diff
@@ -1249,17 +1249,87 @@ int InputProcessor::getObjectItemNum(EnergyPlusData &state,
     return getIDFObjNum(state, ObjType, object_item_num); // if incoming input is idf, then return idf object order
 }
 
+void InputProcessor::lowerRangeCheck(EnergyPlusData &state,
```
Comment (amirroth) [inline src/EnergyPlus/InputProcessing/InputProcessor.cc:1252]: General comment about functions like these. You are passing 9 parameters to this function, 8 of which are only needed in the case of an error, which is presumably rare. This is MCCS (Make the Common Case Slow) 101. If you want to make a function out of the error message code itself, okay. But do the range check at the call site and only call that function if there is an actual error.

File: src/EnergyPlus/InputProcessing/InputProcessor.cc [hunk 9]
```diff
@@ -1249,17 +1249,87 @@ int InputProcessor::getObjectItemNum(EnergyPlusData &state,
...
+            } else {
+                ShowSevereError(state, Message1);
+                ShowContinueError(state, Message2);
+                ErrorsFound = true;
+            }
+        }
+    }
+}
+
 void InputProcessor::rangeCheck(EnergyPlusData &state,
```
Comment (amirroth) [inline src/EnergyPlus/InputProcessing/InputProcessor.cc:1322]: Same comment as above. Understand that this is not part of this PR, but if I see something I'm going to say something.

File: src/EnergyPlus/ScheduleManager.cc [hunk 10]
```diff
@@ -4258,11 +4258,11 @@ namespace ScheduleManager {
...
 
     bool CheckDayScheduleValueMinMax(EnergyPlusData &state,
-                                     int const ScheduleIndex,        // Which Day Schedule being tested
-                                     Real64 const Minimum,           // Minimum desired value
-                                     std::string const &MinString,   // Minimum indicator ('>', '>=')
-                                     Optional<Real64 const> Maximum, // Maximum desired value
-                                     Optional_string_const MaxString // Maximum indicator ('<', ',=')
+                                     int const ScheduleIndex, // Which Day Schedule being tested
+                                     Real64 const Minimum,    // Minimum desired value
+                                     bool const exclusiveMin, // Minimum indicator ('>', '>=')
```
Comment (amirroth) [inline src/EnergyPlus/ScheduleManager.cc:4263]: 👍

File: src/EnergyPlus/WeatherManager.cc [hunk 11]
```diff
@@ -2634,29 +2634,25 @@ namespace WeatherManager {
...
-                                                                              "Direct Radiation",
-                                                                              "WeatherFile",
-                                                                              "Severe",
-                                                                              ">= 0",
-                                                                              (DirectRad >= 0.0),
-                                                                              _,
-                                                                              _,
-                                                                              _,
-                                                                              state.dataEnvrn->WeatherFileLocationTitle);
+                        state.dataInputProcessing->inputProcessor->lowerRangeCheck(state,
```
Comment (amirroth) [inline src/EnergyPlus/WeatherManager.cc:2637]: See comment above about `rangeCheck` function.

File: src/EnergyPlus/DXCoils.cc (hunk 5 above)
Comment (jasondegraw) [reply src/EnergyPlus/DXCoils.cc:15722]: These functions should hopefully only get called during the input phase, but (a) good point and (b) I agree.

File: src/EnergyPlus/InputProcessing/InputProcessor.cc (hunk 8 above)
Comment (jasondegraw) [reply src/EnergyPlus/InputProcessing/InputProcessor.cc:1252]: I went ahead and refactored the function that got split into this one and the other one below, but this only gets used in WeatherManager regardless of whether it was intended to be a general use function. That's another argument for it to go.

Comment (amirroth) [reply src/EnergyPlus/InputProcessing/InputProcessor.cc:1252]: Yeah. More important than this specific function, however, is the general lesson and pattern. One of the tenets of high-performance (or at least not terrible-performance) programming is "Make the Common Case Fast". Setting up parameters takes time. Don't take that time if those parameters will only be used infrequently, e.g., in the case of an error. Another example I see quite a bit of is functions that have many options that tweak their behavior in different ways. Yes, you are "reusing code" and that's good, but even good things can be taken too far. If those options are rarely used, then you're slowing down the more frequent instances of the function.

These "rules of thumb" are obvious once someone points them out to you. And they are invaluable. I learned this one in my second year graduate computer architecture class. MCCF is *the* rule of thumb for computer hardware design, and it applies to software just as well.

**Comment by Myoldmopar** [comment]
@jasondegraw the last PR just caused a conflict in one file.  I'm happy to resolve if you want me to just let me know.

**Comment by jasondegraw** [comment]
@Myoldmopar Yes, that would be great if you could do that. This should be ready for review, when I last checked it was all good but it may have fallen behind develop a bit.

**Comment by Myoldmopar** [comment]
Conflicts resolved, everything builds and passes fine.  Carry on...

**Comment by jasondegraw** [comment]
@Myoldmopar I think we want to get this in now and get out of the way of some of the other work that is going on, who should do review of this PR?

**Review by mitchute — APPROVED** [review]
Great to have these cleaned up and some of the messiness that came along with these optionals gone. 

This all looks OK to me. Nothing is jumping out as controversial, just a nice set of cleanup changes. I think this can merge right away if CI comes back green.

**Review by Myoldmopar — APPROVED** [review]
Yeah, this looks awesome.  I have no issues with this.

**Comment by Myoldmopar** [comment]
I'm not sure why Windows didn't come back yet, but I'm not going to hold it up.  It was clean ahead of the merge from develop.  Merging now, thanks @jasondegraw
